### PR TITLE
Lsu allow spec load exec

### PR DESCRIPTION
**Pull request by h0lyalg0rithm**
This PR introduces the following
- Allows loads to perform Virtual address to Physical address translation before older loads complete.
  This can be configured using a parameter in the LSU
- Allow non blocking cache lookup requests. Depends on #91 for the cache to support this feature
- Implements Ready queue - Simulator only structure to speed up instruction lookup in the LSU.
- Mitigates data hazards introduced by instructions running out of order
- The length of the different stages of the LSU pipeline can be configured through the parameters.

### CODE REVIEW

**Review by arupc — COMMENTED** [review]
Can we add some documentation on what are the purpose of various in and out ports of LSU, at least between LSU & MMU and LSU & Dache

LSU / MMU interface has 3 input ports and 1 output:
* output:
   - out_mmu_lookup_req
 * inputs:
   - in_mmu_lookup_req
   - in_mmu_lookup_ack
   - in_mmu_free_req
  
Similary, LSU / Dache interface has 3 input ports and 1 output port:
* output:
   - out_cache_lookup_req
 * inputs:
   - in_cache_lookup_req
   - in_cache_lookup_ack
   - in_cache_free_req

File: core/LSU.cpp
Comment (arupc) [inline core/LSU.cpp:731]: Please add documentation (even if just one line) stating what each method is doing.

File: core/LSU.hpp
Comment (arupc) [inline core/LSU.hpp:364]: what do we need both `instPresentInQueues` and `instInQueue`? 

Also, by "queue", do we mean just the issue queue (as opposed to load queue, store queue)? If so, may be change the comment to reflect that.

File: core/LSU.hpp
Comment (arupc) [inline core/LSU.hpp:377]: Please add one line comment describing the purpose of each the above methods

File: core/LSU.cpp
Comment (arupc) [inline core/LSU.cpp:156]: Is this condition ever going to be false? If this condition is false, then does it mean, dispatch sent an instruction which is already there in the ldst issue queue?

File: core/LSU.hpp
Comment (arupc) [inline core/LSU.hpp:380]: All private methods should be  suffixed with underscore, following the convention of the rest of the code and to enhance the readability of the code

**Review by arupc — COMMENTED** [review]
Can you please explain what these priorities / states mean and when they would occur? Not sure I understand what is meant by 'mss ack' and 'another oustanding miss finish' 

```
                HIGHEST = 0,
                __FIRST = HIGHEST,
                CACHE_RELOAD,   // Receive mss ack, waiting for cache re-access
                CACHE_PENDING,  // Wait for another outstanding miss finish
                MMU_RELOAD,     // Receive for mss ack, waiting for mmu re-access
                MMU_PENDING,    // Wait for another outstanding miss finish
                NEW_DISP,       // Wait for new issue
                LOWEST,
```

Does `NEW_DISP` simply mean dispatched and waiting to be issued?

**Review by arupc — COMMENTED** [review]
One drawback of having a enumeration of IssuePriority is that it _somewhat_ ties us to having one  issue priority policy. If we want to allow multiple issue priority policies as a paramter, it may be difficult to with IssuePriority enumerator.

For that, we should keep the CACHE/MMU_RELOAD/PENDING info as states of the LoadStoreInfo class and allow different policies to decide which one to issue.

This can be done in a different PR after this is merged in.

File: core/LSU.cpp
Comment (h0lyalg0rithm) [reply core/LSU.cpp:731]: Yes I will add them

**Review by knute-sifive — DISMISSED** [review]
This is a great start.  A few items here and there to tackle, obviously

File: core/LSU.hpp
Comment (h0lyalg0rithm) [reply core/LSU.hpp:295]: @knute-sifive I use these variables as an alias for the pipeline stage. In the constructor I would replace them based on the length of the pipeline the user sets as a parameter.

File: core/LSU.hpp
Comment (h0lyalg0rithm) [reply core/LSU.hpp:59]: I designed the parameters to be in setup in a ldst_pipeline agnostic way
By default these are the mappings

| Stage         | Pipeline Alias     | Default |
|--------------|-----------|----------- |
| ADDRESS_GENERATION | 0      | 0 |
| MMU_LOOKUP      |  mmu_lookup_stage | 1 |
| CACHE_LOOKUP      |  mmu_lookup_stage + cache_lookup_stage | 2 |
| CACHE_READ      |  mmu_lookup_stage + cache_lookup_stage + cache_read_stage | 3 |
| COMPLETE      |  mmu_lookup_stage + cache_lookup_stage + cache_read_stage + 1 | 4 |

This would ensure that the stages are in the order without checking for the overlaps as the arithmetic would handle it.I should add validation to ensure that the length parameters are atleast 1.

example:
```
mmu_lookup_stage_length => 1
cache_lookup_stage_length => 2
cache_read_stage_length =>  2
```
| Stage         | Pipeline Alias     | Value |
|--------------|-----------| -----------|
| ADDRESS_GENERATION | 0      | 0  |
| MMU_LOOKUP      |  mmu_lookup_stage | 1|
| CACHE_LOOKUP      |  mmu_lookup_stage + cache_lookup_stage | 3|
| CACHE_READ      |  mmu_lookup_stage + cache_lookup_stage + cache_read_stage | 5 |
| COMPLETE      |  mmu_lookup_stage + cache_lookup_stage + cache_read_stage + 1 | 6 |

File: core/LSU.cpp
Comment (h0lyalg0rithm) [reply core/LSU.cpp:156]: Yes this condition can be false. Currently there is a callback that is registered in the same method.
One for one of operands and other for the data 
https://github.com/riscv-software-src/riscv-perf-model/pull/92/files/11626bef0bcdc815dd6cc9b37dbe7f2d6d558858#diff-f42bfd79bf07a70c3e64bbe0c640480fd5ee792c0c0465f138d9bad9219959a0R183

https://github.com/riscv-software-src/riscv-perf-model/pull/92/files/11626bef0bcdc815dd6cc9b37dbe7f2d6d558858#diff-f42bfd79bf07a70c3e64bbe0c640480fd5ee792c0c0465f138d9bad9219959a0R170

**Comment by h0lyalg0rithm** [comment]
> Can you please explain what these priorities / states mean and when they would occur? Not sure I understand what is meant by 'mss ack' and 'another oustanding miss finish'
> 
> ```
>                 HIGHEST = 0,
>                 __FIRST = HIGHEST,
>                 CACHE_RELOAD,   // Receive mss ack, waiting for cache re-access
>                 CACHE_PENDING,  // Wait for another outstanding miss finish
>                 MMU_RELOAD,     // Receive for mss ack, waiting for mmu re-access
>                 MMU_PENDING,    // Wait for another outstanding miss finish
>                 NEW_DISP,       // Wait for new issue
>                 LOWEST,
> ```
> 
> Does `NEW_DISP` simply mean dispatched and waiting to be issued?
The priority values are used to select the instruction that needs to be issued.
NEW_DISP is the priority of an instruction which has its operands ready(depends on the spec_load_exec) and is ready to be issued.
CACHE_RELOAD on the other hand is the priority placed on an instruction which had previously had a cache miss and now the cache is ready so it wake up the instruction by setting its priority to CACHE_RELOAD. As the priority is higher it is selected to be issued.

File: core/LSU.hpp
Comment (knute-sifive) [reply core/LSU.hpp:59]: Ok, I can buy that for a dollar 💵 (sorry, bad movie reference).

File: core/LSU.hpp
Comment (h0lyalg0rithm) [reply core/LSU.hpp:364]: Yes I should rename it, its made sense when I had multiple queues

**Comment by arupc** [comment]
> > Can you please explain what these priorities / states mean and when they would occur? Not sure I understand what is meant by 'mss ack' and 'another oustanding miss finish'
> > ```
> >                 HIGHEST = 0,
> >                 __FIRST = HIGHEST,
> >                 CACHE_RELOAD,   // Receive mss ack, waiting for cache re-access
> >                 CACHE_PENDING,  // Wait for another outstanding miss finish
> >                 MMU_RELOAD,     // Receive for mss ack, waiting for mmu re-access
> >                 MMU_PENDING,    // Wait for another outstanding miss finish
> >                 NEW_DISP,       // Wait for new issue
> >                 LOWEST,
> > ```
> > 
> > 
> > Does `NEW_DISP` simply mean dispatched and waiting to be issued?
> 
> The priority values are used to select the instruction that needs to be issued. NEW_DISP is the priority of an instruction which has its operands ready(depends on the spec_load_exec) and is ready to be issued. CACHE_RELOAD on the other hand is the priority placed on an instruction which had previously had a cache miss and now the cache is ready so it wake up the instruction by setting its priority to CACHE_RELOAD. As the priority is higher it is selected to be issued.

**Comment by arupc** [comment]
> > > Can you please explain what these priorities / states mean and when they would occur? Not sure I understand what is meant by 'mss ack' and 'another oustanding miss finish'
> > > ```
> > >                 HIGHEST = 0,
> > >                 __FIRST = HIGHEST,
> > >                 CACHE_RELOAD,   // Receive mss ack, waiting for cache re-access
> > >                 CACHE_PENDING,  // Wait for another outstanding miss finish
> > >                 MMU_RELOAD,     // Receive for mss ack, waiting for mmu re-access
> > >                 MMU_PENDING,    // Wait for another outstanding miss finish
> > >                 NEW_DISP,       // Wait for new issue
> > >                 LOWEST,
> > > ```
> > > 
> > > 
> > > Does `NEW_DISP` simply mean dispatched and waiting to be issued?
> > 
> > 
> > The priority values are used to select the instruction that needs to be issued. NEW_DISP is the priority of an instruction which has its operands ready(depends on the spec_load_exec) and is ready to be issued. CACHE_RELOAD on the other hand is the priority placed on an instruction which had previously had a cache miss and now the cache is ready so it wake up the instruction by setting its priority to CACHE_RELOAD. As the priority is higher it is selected to be issued.

How would we implement an issue policy which prioritizes requests with MMU pending over request with CACHE pending?

How about instead of IssuePriority enum, we keep something equivalent to bitmasks, specifying for each instance of LoadStoreInstInfo, whether MMU_LOOKUP is needed and has been satisfied, CACHE_LOOKUP is needed and has been satisfied.  

This would enable implementation of various issue policies, without changing the enum. But this can be done as apart of another MR

File: core/LSU.cpp
Comment (arupc) [inline core/LSU.cpp:223]: should it be an assert instead of if(....), since we just updated issue priority? Does that mean we are guaranteed to find one inst that is ready?

File: core/LSU.cpp
Comment (arupc) [inline core/LSU.cpp:251]: why are we checking if allow_speculative_load_exec is true here? Even if speculative load exec is false, a load or a store can miss in MMU or cache and result in replay and hence, would allocate into the replay buffer during issue.

File: core/LSU.cpp
Comment (arupc) [inline core/LSU.cpp:264]: same comment as above

File: core/LSU.cpp
Comment (arupc) [inline core/LSU.cpp:28]: Can you please comment why we need +2?

File: core/LSU.hpp
Comment (h0lyalg0rithm) [reply core/LSU.hpp]: I made the change, Previously I was  changing them through a function call switch to constructor initialization

File: core/LSU.cpp
Comment (h0lyalg0rithm) [reply core/LSU.cpp:222]: Yes isReadytoIssueInsts_ finds a instruction that can be issued.Yes if it returns true it should issue.

File: core/LSU.cpp
Comment (h0lyalg0rithm) [reply core/LSU.cpp:250]: We use the replay buffer only when speculative_load_exec is true. I can remove the check since replay_buffer would default initialize to the size of the load store queue

File: core/LSU.cpp
Comment (h0lyalg0rithm) [reply core/LSU.cpp]: Its for the address_generation and completion stage length, currently they are of length 1 + 1

File: core/LSU.cpp
Comment (h0lyalg0rithm) [reply core/LSU.cpp]: Yeah my bad should get rid of all them now

File: core/LSU.cpp
Comment (knute-sifive) [inline core/LSU.cpp:883]: Remove the `inline` keyword

File: core/LSU.cpp
Comment (knute-sifive) [inline core/LSU.cpp:911]: Remove the `inline` keyword

File: core/LSU.cpp
Comment (knute-sifive) [inline core/LSU.cpp:920]: If you add a variable to the `LoadStoreInstInfo` that captures where it is, you can double check to make sure it's not already enqueued:
```
    sparta_assert(false == inst_info_ptr->isInIssueQueue(), "Instruction already in queue: " << inst_info_ptr);
    inst_info_ptr->setInIssueQueue(true);
```

This removes the expensive `find_if`

File: core/LSU.cpp
Comment (knute-sifive) [inline core/LSU.cpp:712]: Why is the `InstPtr` being passed here?  Shouldn't it be a `LoadStoreInfo` type?  If you change this (strong suggestion you do), you can stash the replay iterator in the `LoadStoreInfo` and perform one operation:
```
replay_buffer_->erase(load_store_info_->getReplayIterator());
```

File: core/LSU.cpp
Comment (knute-sifive) [inline core/LSU.cpp:947]: Remove `allow_speculative_load_exec_` from the condition.  You've already checked for this.

**Comment by h0lyalg0rithm** [comment]
Thanks just fixed the conflicts @klingaard

**Comment by klingaard** [comment]
@h0lyalg0rithm can you provide an update on where you are with this PR?  Are you waiting on our approval now?

**Comment by h0lyalg0rithm** [comment]
@klingaard Yes I am waiting on the approval

**Comment by klingaard** [comment]
Ok, please resolve all conversations that you feel are addressed.  Thanks!

**Comment by knute-sifive** [comment]
Did you test this with Kunal L2Cache stuff?  Does it work?

File: test/sim/CMakeLists.txt
Comment (h0lyalg0rithm) [reply test/sim/CMakeLists.txt:55]: It must be a mistake let me take a look

**Comment by h0lyalg0rithm** [comment]
> Did you test this with Kunal L2Cache stuff? Does it work?

I havent been able to test it out, I tried to keep the PR in sync with master

**Comment by h0lyalg0rithm** [comment]
@klingaard I rebuilt the project(cleant cache, cmake build folder) I ran both coremark and drhystone and both of them ran successfully.
I have the logs from valgrind but there were tons of errors but not sure if they are false positive

` ERROR SUMMARY: 10000034 errors from 41 contexts`

**Comment by klingaard** [comment]
I'll take a look this evening.

File: core/LSU.cpp
Comment (klingaard) [inline core/LSU.cpp:272]: This is the valgrind error I found.  After the `pop` operation, `win_ptr` is no longer valid (although the memory is still on the stack).  

What is happening:
1. You call `arbitrateInstIssue` which returns the Ptr by reference and you make a `const` reference to be used locally
2. Then, here, you `pop` it and now that pointer is nullified.

Suggestions on fixing:
Have `arbitrateInstIssue` return the pointer by value AND perform the `pop` _in that function_.

Comment (klingaard) [reply core/LSU.cpp]: Looks good.  After making another fix to DCache.hpp, valgrind is clean.  Good stuff!

**Comment by knute-sifive** [comment]
I forgot one more thing.  😄  Can you run run clang format on the LSU since you're the last person to make major changes to it?

**Comment by klingaard** [comment]
@h0lyalg0rithm, I hijacked your PR to update the names of the CI regressions (and merged with master).  FYI...

**Comment by h0lyalg0rithm** [comment]
@klingaard I ran the formatter on most of the files I touched.I left some like `Inst.hpp` and `sim/Olympia.cpp`. 
I found an duplicate config in the clang format file. Could you please check I removed the correct one

File: .clang-format
Comment (h0lyalg0rithm) [inline .clang-format:141]: The same key is present here https://github.com/riscv-software-src/riscv-perf-model/pull/92/commits/44291791647e5e562f4a1820be98555cb7f81668#diff-1026e0038b722990204a42bed8a6f7c0ec2302aa79e3fad1959d62ba968edfa2L192

**Comment by klingaard** [comment]
Thanks for running the formatter.  One last, last thing, promise!  Can you update the Description field on what this PR did -- the changes it implemented and new expected behaviors.  That will be the final commit message.

File: core/LSU.hpp
Comment (klingaard) [inline core/LSU.hpp:55]: @h0lyalg0rithm is this parameter's boolean meaning inverted?   I'm looking at the code and if this is set to `true`, then the LS will NOT allow a load to bypass an older store.

Comment (h0lyalg0rithm) [reply core/LSU.hpp:55]: @klingaard This parameter by default is set to true, so it should allow a load to complete before a store

**Pull request merged**
